Unity-Technologies/ScriptableRenderPipeline
# Nested sub graphs

### Purpose of this PR
This PR enables nesting of sub graphs. It should also improve on the stability of sub graphs. We now bake everything we need to know about a sub graph at import time, such that Sub Graph Node no longer needs any access to the referenced graph. Likewise things like shader stage capability are also computed at import time. 

I also made some modifications to `GraphData.ValidateGraph`. Previously each `ValidateNode` call would recursively call `ValidateNode` on all children, meaning that each node would end up having `ValidateNode` called many times. This caused my error checking code in `SubGraphNode` to spam messages because it was being called multiple times. `GraphData.ValidateGraph` now ensures that all children of a node has had `ValidateNode` called before calling it on the parent node, and so each node doesn't have to call into child nodes anymore.

---
### Testing status
**Katana Tests**: On the way https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?unity_branch=trunk&ScriptableRenderLoop_branch=sg%2Fnested-subgraphs&automation-tools_branch=master

**Manual Tests**: What did you do?
- [x] ~Checked new UI names with UX convention~ No new UI
- [x] ~Tested UI multi-edition~ (Not relevant) + Undo/Redo
- Verify that dependency tracking works for main and sub graphs using nested sub graphs.
- Verified that the recursion check works and triggers errors in the console.
- Verified that adding in a recursion in a nested sub graph, and then removing that recursion, causes the main graph to go pink and then back to working.

**Automated Tests**: Nothing new.

---
### Overall Product Risks
**Technical Risk**: Medium

**Halo Effect**: Low

---
### Comments to reviewers
Most of the action happens in SubGraphDatabaseImporter and SubGraphNode. The former class is an importer for a new asset type, for which only 1 asset at a fixed path can exist. I've set SubGraphImporter to have an import queue offset of 1, SubGraphDatabaseImporter to have an offset of 2, and finally ShaderGraphImporter to have an offset of 3. This means that the new importer essentially acts as a callback in between sub graphs and main graphs. The SubGraphImporter now only imports a timestamp, which I use in SubGraphDatabaseImporter to determine whether a sub graph (and all sub graphs that are using it) should be re-processed.

Please do try to stress this as much as you can by nesting sub graphs many times and using them in lots of places.
